kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -40,25 +43,27 @@ pub trait AnySocket {
 
 pub type BoundTcpSocket<E> = BoundSocket<TcpSocket, E>;
 pub type BoundUdpSocket<E> = BoundSocket<UdpSocket, E>;
+pub type BoundRawSocket<E> = BoundSocket<RawSocket, E>;
 
-/// Common states shared by [`BoundTcpSocketInner`] and [`BoundUdpSocketInner`].
+/// Common states shared by [`BoundTcpSocketInner`] , [`BoundUdpSocketInner`] and [`BoundRawSocketInner`].
 pub struct BoundSocketInner<T, E> {
     iface: Arc<dyn Iface<E>>,
     port: u16,
     socket: T,
     observer: RwLock<Weak<dyn SocketEventObserver>, LocalIrqDisabled>,
     events: AtomicU8,
     next_poll_at_ms: AtomicU64,
+    ip_protocol: IpProtocol,
```

Comment:
The protocol field is used to filter out network packets suitable for the socket. This field should be placed in the socketBg structure. However, at this time, this field is redundant for tcp and udp, as you said, and the port field is redundant for raw sockets. Using one structure to abstract all socket types is always not perfect, and there are more or less redundant fields. Maybe we should just accept this fact and allow these redundant fields to exist? 
If there is a better way, please tell me, thanks.

Reply:
> The protocol field is used to filter out network packets suitable for the socket. This field should be placed in the socketBg structure. However, at this time, as you said, this field is redundant for tcp and udp, and the port field is redundant for raw sockets. Using one structure to abstract all socket types is always not perfect, and there are more or less redundant fields. Maybe we should just accept this fact and allow these redundant fields to exist?
> If there is a better way, please tell me, thanks.


The raw socket structure in smoltcp has a protocol field, which temporarily solves my problem. However, the port field is redundant for raw sockets, which may be acceptable？

Reply:
> However, the port field is redundant for raw sockets, which may be acceptable？

I admit it's a pain to deal with this problem.

In fact, even if we don't consider raw sockets, TCP/UDP sockets also have different meanings in the port field.
 - In other words, TCP ports and UDP ports should be assigned independently, i.e., there can be a TCP socket and a UDP socket that have the same port assignment, but there should be no conflict because they are different protocols, but this is not correctly implemented in Asterinas for now.

For raw sockets, it has no port assignment, so you may want to use something like:
```rust
struct BoundPort {
    iface: Arc<dyn Iface>,
    port: Option<NonZeroU16>,
}
```
Then, just let `port` be `None`.

We can consider the refactoring of `BoundPort` as future work.